src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs

```diff
@@ -813,24 +813,43 @@ private static unsafe void StelemRef_Helper(ref object element, MethodTable* ele
             }
         }
 
+        private static unsafe ref object ThrowArrayMismatchException(Array array)
+        {
+            // Throw the array type mismatch exception defined by the classlib, using the input array's MethodTable*
+            // to find the correct classlib.
+            throw array.GetMethodTable()->GetClasslibException(ExceptionIDs.ArrayTypeMismatch);
```

Comment:
Nit: would be nice to include the element type in the exception - otherwise it's like saying "You created array with wrong element type... but we won't tell you which one... "
(But maybe this only happens with invalid IL and thus it doesn't matter much)

Reply:
We have a bug somewhere. In the old .NET Native times we were not guaranteed we could get a name. Now we could. It has some recursion dangers though (what if reflection implementation also needs to throw this while getting a name).